process/block/preprocess/basePreProcess.go

```diff
@@ -324,13 +327,20 @@ func (bpp *basePreProcess) computeGasConsumed(
 	if bpp.shardCoordinator.SelfId() == senderShardId {
 		gasConsumedByTxInSelfShard = gasConsumedByTxInSenderShard
 
-		if *gasConsumedByMiniBlockInReceiverShard+gasConsumedByTxInReceiverShard > bpp.economicsFee.MaxGasLimitPerBlock(bpp.shardCoordinator.SelfId()) {
+		if gasConsumedByTxInReceiverShard > bpp.economicsFee.MaxGasLimitPerMiniBlock(process.ShardIDWithSafeMaxGasLimit) {
+			return process.ErrMaxGasLimitPerOneTxInReceiverShardIsReached
+		}
+
+		//TODO: MaxGasLimitPerBlock method should have parameter process.ShardIDWithSafeMaxGasLimit instead bpp.shardCoordinator.SelfId(),
+		//as if this mini block will be created by meta for shards, it could be created with 15 bil. instead 1.5 bil. gas,
+		//and the receiver shard will be stuck
+		if *gasConsumedByMiniBlockInReceiverShard+gasConsumedByTxInReceiverShard > bpp.economicsFee.MaxGasLimitPerBlock(process.ShardIDWithSafeMaxGasLimit) {
```

Comment:
you change every check of maxGasLimitPerBlock with ShardIDWithSafeMaxGasLimit - can we refactor this function, in order to be save in all the cases ? If we have a set of transaction to meta with a lot of refunded gas - this might give some strange problems - and the shard would be capable of processing it.

Reply:
This replace is not done in all cases, only when I am the sender and I will create something for receiver, either if the receiver is myself or other shard. The second replace is done when the sender is not me and the check is done with ShardIDWithSafeMaxGasLimit only for the sender side. If you check the L345 below in this method you will see that the check for the gas consumed in self shard is still done with MaxGasLimitPerBlock(bpp.shardCoordinator.SelfId()), so the meta will have the chance to include txs until 15 bil. gas if there are many mbs to himself from many shards.